locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsByonLocationResolver.java

```diff
@@ -61,129 +56,140 @@
+    @Override
+    protected SpecParser getSpecParser() {
+        return new AbstractLocationResolver.SpecParser(getPrefix()).setExampleUsage("\"jcloudsByon(provider='aws-ec2',region='us-east-1',hosts='i-12345678,i-90123456')\"");
+    }
+    
+    @Override
+    protected ConfigBag extractConfig(Map<?,?> locationFlags, String spec, final LocationRegistry registry) {
+        ConfigBag config = super.extractConfig(locationFlags, spec, registry);
 
-        String providerOrApi = argsMap.containsKey("provider") ? argsMap.get("provider") : (String)locationFlags.get("provider");
+        String providerOrApi = (String) config.getStringKey("provider");
+        String regionName = (String) config.getStringKey("region");
+        String endpoint = (String) config.getStringKey("endpoint");
+        String namedLocation = (String) config.get(LocationInternal.NAMED_SPEC_NAME);
+        config.remove(LocationInternal.NAMED_SPEC_NAME.getName());
 
-        String regionName = argsMap.containsKey("region") ? argsMap.get("region") : (String)locationFlags.get("region");
-        
-        String endpoint = argsMap.containsKey("endpoint") ? argsMap.get("endpoint") : (String)locationFlags.get("endpoint");
-        
-        String name = argsMap.containsKey("name") ? argsMap.get("name") : (String)locationFlags.get("name");
+        Object hosts = config.getStringKey("hosts");
+        config.remove("hosts");
 
-        String user = argsMap.containsKey("user") ? argsMap.get("user") : (String)locationFlags.get("user");
+        Map jcloudsProperties = new JcloudsPropertiesFromBrooklynProperties().getJcloudsProperties(providerOrApi, regionName, namedLocation, config.getAllConfig());
+        config.putIfAbsent(jcloudsProperties);
 
-        String privateKeyFile = argsMap.containsKey("privateKeyFile") ? argsMap.get("privateKeyFile") : (String)locationFlags.get("privateKeyFile");
-        
-        String hosts = argsMap.get("hosts");
-        
         if (Strings.isEmpty(providerOrApi)) {
             throw new IllegalArgumentException("Invalid location '"+spec+"'; provider must be defined");
         }
-        if (hosts == null || hosts.isEmpty()) {
+        if (hosts == null || (hosts instanceof String && Strings.isBlank((String)hosts))) {
             throw new IllegalArgumentException("Invalid location '"+spec+"'; at least one host must be defined");
         }
-        if (argsMap.containsKey("name") && (Strings.isEmpty(name))) {
-            throw new IllegalArgumentException("Invalid location '"+spec+"'; if name supplied then value must be non-empty");
+
+        final String jcloudsSpec = "jclouds:"+providerOrApi + (regionName != null ? ":"+regionName : "") + (endpoint != null ? ":"+endpoint : "");
+        final Maybe<LocationSpec<? extends Location>> jcloudsLocationSpec = registry.getLocationSpec(jcloudsSpec, config.getAllConfig());
+        if (jcloudsLocationSpec.isAbsent()) {
```

Comment:
Changing to `if (jcloudsLocationSpec.isAbsentOrNull())`.

I really don't get why we're using `Maybe` here, rather than `Optional`. The contract would be far more sensible if `getLocationSpec` guaranteed to never return `Maybe.of(null)`! But we can't change the return type without breaking backwards compatibility :-(